shared/src/main/scala/mlscript/NewLexer.scala

```diff
@@ -31,9 +31,11 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
     c === '0' || c === '1'
   def isDigit(c: Char): Bool =
     c >= '0' && c <= '9'
-  def matches(i: Int, syntax: Str): Bool =
-    i + syntax.length <= length && bytes.slice(i, i + syntax.length).mkString === syntax
-  def isQuasiquote(i: Int): Bool = matches(i, QQCode.prefix + '"') // code"
+  def matches(i: Int, syntax: Str, start: Int): Bool =
+    if (start >= syntax.length) true
+    else if (i + start < length && bytes(i + start) === syntax(start)) matches(i, syntax, start + 1)
+    else false
```

Comment:
Pro-tip: if you end up writing `if ... then true` and `else false`, something has gone wrong ;^)